group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -761,6 +777,31 @@ public ClassicGroup classicGroup(
         }
     }
 
+    public boolean validateOnlineUpgrade(ClassicGroup classicGroup) {
+        return ConsumerGroupMigrationPolicy.isUpgradeEnabled(consumerGroupMigrationPolicy) &&
+            !classicGroup.isInState(DEAD) &&
+            ConsumerProtocol.PROTOCOL_TYPE.equals(classicGroup.protocolType().orElse(null)) &&
+            classicGroup.size() <= consumerGroupMaxSize;
+    }
+
+    ConsumerGroup convertToConsumerGroup(ClassicGroup classicGroup, List<Record> records) {
+        classicGroup.completeAllJoinFutures(Errors.REBALANCE_IN_PROGRESS);
+        classicGroup.completeAllSyncFutures(Errors.REBALANCE_IN_PROGRESS);
+
+        createGroupTombstoneRecords(classicGroup, records);
+        ConsumerGroup consumerGroup = new ConsumerGroup(snapshotRegistry, classicGroup.groupId(), metrics);
+        classicGroup.convertToConsumerGroup(consumerGroup, records, metadataImage.topics());
```

Comment:
I was wondering whether it would make more sense the other way around and have something like `ConsumerGroup.fromClassicGroup(....)`. I guess that it does not really matter in the end.